cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -809,6 +809,475 @@ struct IsUpperAscii : CharacterPredicateAscii<IsUpperAscii> {
+            *(*output_string_data)++ = *i++;
+            (*string_output_offset)++;
+          }
+          // 'finish' the string by writing the offset
+          *(*output_string_offsets)++ = *string_output_offset;
+          (*string_output_count)++;
+          // jump of the part we cut out
+          i = separator_end;
+          max_splits--;
+        } else {
+          // if we cannot find a separator, we're done
+          break;
+        }
+      }
+      // copy bytes after the pattern
+      while (i < end) {
+        *(*output_string_data)++ = *i++;
+        (*string_output_offset)++;
+      }
+      // and write out the trailing part (can be an empty string)
+      *(*output_string_offsets)++ = *string_output_offset;
+      (*string_output_count)++;
+    }
+  }
+  static Status CheckOptions(const Options& options) { return Status::OK(); }
+  static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    EnsureLookupTablesFilled();  // only needed for unicode
+    Options options = State::Get(ctx);
+    KERNEL_RETURN_IF_ERROR(ctx, Derived::CheckOptions(options));
+
+    if (batch[0].kind() == Datum::ARRAY) {
+      const ArrayData& input = *batch[0].array();
+      ArrayType input_boxed(batch[0].array());
+      ArrayData* output_list_data = out->mutable_array();
+
+      offset_type input_nbytes = input_boxed.total_values_length();
+      offset_type input_nstrings = static_cast<offset_type>(input.length);
+
+      offset_type output_nbytes_max = input_nbytes;
+      int64_t output_nstrings_max = Derived::CalculateMaxSplits(input_boxed, options);
+      if (output_nstrings_max > std::numeric_limits<offset_type>::max()) {
+        ctx->SetStatus(
+            Status::CapacityError("Result might not fit in a 32bit list or string array, "
+                                  "convert to large_utf8"));
+        return;
+      }
+
+      // Why is the offset buffer not preallocated?
+      KERNEL_ASSIGN_OR_RAISE(output_list_data->buffers[1], ctx,
+                             ctx->Allocate((input_nstrings + 1) * sizeof(offset_type)));
+      offset_type* output_list_offsets =
+          output_list_data->GetMutableValues<offset_type>(1);
+
+      // allocate output string array data
+      // this is a bit low level, should we use a builder?
+      KERNEL_ASSIGN_OR_RAISE(auto buffer_output_string_data, ctx,
+                             ctx->Allocate((output_nbytes_max)));
+      KERNEL_ASSIGN_OR_RAISE(auto buffer_output_string_offsets, ctx,
+                             ctx->Allocate(output_nstrings_max * sizeof(offset_type)));
+
+      offset_type* output_string_offsets =
+          reinterpret_cast<offset_type*>(buffer_output_string_offsets->mutable_data());
+      uint8_t* output_string_data = buffer_output_string_data->mutable_data();
+      // TypedBufferBuilder<bool> null_builder;  // TODO: should we pass a mem poool?
+      // KERNEL_RETURN_IF_ERROR(ctx, null_builder.Reserve(output_nstrings_max));
```

Comment:
Why is all this commented out? Is this PR unfinished?

Reply:
This was related to the issue related to the test that is commented out (and described in the opening post). Reading @jorisvandenbossche explanation I think I did it correctly, and this code can be ignored. I think there is an issue with equality testing and null values in lists (as described above).

Reply:
If there is such an issue, it would be good to open a dedicated JIRA with a reproducer. Let's not bury this in the comments of a string compute PR :-)

Reply:
done https://issues.apache.org/jira/browse/ARROW-10208